File: regions/system/library/go/building-blocks/inmemory_secretstore.go
package buildingblocks

import (
	"context"
	"fmt"
	"sync"
)

// Compile-time interface compliance check.
var _ SecretStore = (*InMemorySecretStore)(nil)

// InMemorySecretStore is an in-memory implementation of SecretStore for testing.
type InMemorySecretStore struct {
	mu      sync.RWMutex
	secrets map[string]*Secret
	status  ComponentStatus
}

// NewInMemorySecretStore creates a new InMemorySecretStore.
func NewInMemorySecretStore() *InMemorySecretStore {
	return &InMemorySecretStore{
		secrets: make(map[string]*Secret),
		status:  StatusUninitialized,
	}
}

func (s *InMemorySecretStore) Name() string    { return "inmemory-secretstore" }
func (s *InMemorySecretStore) Version() string { return "1.0.0" }

func (s *InMemorySecretStore) Init(_ context.Context, _ Metadata) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.status = StatusReady
	return nil
}

func (s *InMemorySecretStore) Close(_ context.Context) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.status = StatusClosed
	return nil
}

func (s *InMemorySecretStore) Status(_ context.Context) ComponentStatus {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.status
}

// Get retrieves a secret by key.
func (s *InMemorySecretStore) Get(_ context.Context, key string) (*Secret, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	secret, ok := s.secrets[key]
	if !ok {
		return nil, NewComponentError("inmemory-secretstore", "Get", fmt.Sprintf("secret %q not found", key), nil)
	}
	return secret, nil
}

// BulkGet retrieves multiple secrets by key.
func (s *InMemorySecretStore) BulkGet(ctx context.Context, keys []string) ([]*Secret, error) {
	results := make([]*Secret, 0, len(keys))
	for _, key := range keys {
		secret, err := s.Get(ctx, key)
		if err != nil {
			return nil, err
		}
		results = append(results, secret)
	}
	return results, nil
}

// Put stores a secret. This is a test helper not part of the SecretStore interface.
func (s *InMemorySecretStore) Put(key, value string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.secrets[key] = &Secret{Key: key, Value: value}
}

File: regions/system/library/typescript/building-blocks/src/inmemory_secretstore.ts
// インメモリシークレットストア実装モジュール。
// テストおよびローカル開発環境向けに、Vault等の外部シークレット管理サービス不要でSecretStoreを提供する。

import { ComponentError } from './errors.js';
import type { ComponentStatus } from './component.js';
import type { SecretValue, SecretStore } from './secretstore.js';

/**
 * インメモリシークレットストア実装クラス。
 * キーと平文値のMapでシークレットを管理する。
 * 本番環境では実際のVault等を使用し、このクラスはテスト専用として扱う。
 */
export class InMemorySecretStore implements SecretStore {
  // コンポーネントの識別名
  readonly name: string;
  // コンポーネント種別（シークレットストアであることを示す）
  readonly componentType = 'secretstore';

  // 現在のコンポーネント状態（未初期化・準備完了・クローズ済み）
  private _status: ComponentStatus = 'uninitialized';
  // キー → 平文値のマッピングを保持するMap
  private readonly _secrets = new Map<string, string>();

  // デフォルト名を指定してインスタンスを生成する
  constructor(name = 'inmemory-secretstore') {
    this.name = name;
  }

  // コンポーネントを初期化し、状態をreadyに変更する
  async init(): Promise<void> { this._status = 'ready'; }

  // コンポーネントをクローズし、全シークレットをクリアして状態をclosedに変更する
  async close(): Promise<void> { this._secrets.clear(); this._status = 'closed'; }

  // 現在のコンポーネント状態を返す
  async status(): Promise<ComponentStatus> { return this._status; }

  // バックエンドを示すメタデータを返す
  metadata(): Record<string, string> { return { backend: 'memory' }; }

  // テスト用: シークレットを設定する。
  put(key: string, value: string): void {
    this._secrets.set(key, value);
  }

  /**
   * 指定キーのシークレットを取得して返す。
   * キーが存在しない場合はComponentErrorをスローする。
   */
  async getSecret(key: string): Promise<SecretValue> {
    const value = this._secrets.get(key);
    // キーが見つからない場合はエラーをスローする
    if (value === undefined) {
      throw new ComponentError(this.name, 'getSecret', `secret "${key}" not found`);
    }
    // キー・値・空のメタデータを含むSecretValueオブジェクトを返す
    return { key, value, metadata: {} };
  }

  /**
   * 複数キーのシークレットをまとめて取得して返す。
   * 内部でgetSecret()を呼び出すため、いずれかのキーが存在しない場合はエラーをスローする。
   */
  async bulkGet(keys: string[]): Promise<Record<string, SecretValue>> {
    const result: Record<string, SecretValue> = {};
    // 各キーのシークレットを順次取得してresultに集約する
    for (const key of keys) {
      result[key] = await this.getSecret(key);
    }
    return result;
  }
}
